retrievalmarket/impl/clientstates/client_fsm.go

```diff
@@ -252,35 +261,68 @@ var ClientEvents = fsm.Events{
 			deal.Message = xerrors.Errorf("writing deal payment: %w", err).Error()
 			return nil
 		}),
-	fsm.Event(rm.ClientEventPaymentSent).
+
+	// Payment was requested, but there was not actually any payment due, so
+	// no payment voucher was actually sent
+	fsm.Event(rm.ClientEventPaymentNotSent).
+		From(rm.DealStatusOngoing).ToJustRecord().
 		From(rm.DealStatusSendFunds).To(rm.DealStatusOngoing).
+		From(rm.DealStatusSendFundsLastPayment).To(rm.DealStatusFinalizing),
+
+	fsm.Event(rm.ClientEventPaymentSent).
+		From(rm.DealStatusOngoing).ToJustRecord().
+		From(rm.DealStatusBlocksComplete).To(rm.DealStatusCheckComplete).
+		From(rm.DealStatusCheckComplete).ToNoChange().
+		FromMany(
+			rm.DealStatusFundsNeeded,
+			rm.DealStatusFundsNeededLastPayment,
+			rm.DealStatusSendFunds).To(rm.DealStatusOngoing).
 		From(rm.DealStatusSendFundsLastPayment).To(rm.DealStatusFinalizing).
-		Action(func(deal *rm.ClientDealState) error {
-			// paymentRequested = 0
-			// fundsSpent = fundsSpent + paymentRequested
-			// if paymentRequested / pricePerByte >= currentInterval
-			// currentInterval = currentInterval + proposal.intervalIncrease
-			// bytesPaidFor = bytesPaidFor + (paymentRequested / pricePerByte)
-			deal.FundsSpent = big.Add(deal.FundsSpent, deal.PaymentRequested)
-
-			paymentForUnsealing := big.Min(deal.PaymentRequested, big.Sub(deal.UnsealPrice, deal.UnsealFundsPaid))
-			deal.UnsealFundsPaid = big.Add(deal.UnsealFundsPaid, paymentForUnsealing)
-
-			// If the price per bytes is zero, we ONLY need to account for the Unsealing payments here.
-			if !deal.PricePerByte.IsZero() {
-				bytesPaidFor := big.Div(big.Sub(deal.PaymentRequested, paymentForUnsealing), deal.PricePerByte).Uint64()
-				if bytesPaidFor >= deal.CurrentInterval {
-					deal.CurrentInterval += deal.DealProposal.PaymentIntervalIncrease
-				}
-				deal.BytesPaidFor += bytesPaidFor
+		Action(func(deal *rm.ClientDealState, voucherAmt abi.TokenAmount) error {
+			// Reduce the payment requested by the amount of funds sent.
+			// Note that it may not be reduced to zero, if a new payment
+			// request came in while this one was being processed.
+			sentAmt := big.Sub(voucherAmt, deal.FundsSpent)
+			deal.PaymentRequested = big.Sub(deal.PaymentRequested, sentAmt)
```

Comment:
What purpose does the `deal.PaymentRequested` variable serve on the client now ?